fdbrpc/AsyncFileChaos.actor.h

```diff
@@ -26,94 +26,103 @@
+			// increment the metric for disk delays
+			if (delayFor > 0.0) {
+				auto res = g_network->global(INetwork::enChaosMetrics);
+				if (res) {
+					ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(res);
+					chaosMetrics->diskDelays++;
+				}
+			}
 		}
+		return delayFor;
 	}
 
 	Future<int> read(void* data, int length, int64_t offset) override {
-		double delay = 0.0;
-		auto res = g_network->global(INetwork::enDiskFailureInjector);
-		if (res)
-			delay = static_cast<DiskFailureInjector*>(res)->getDiskDelay();
-		TraceEvent("AsyncFileChaosRead").detail("ThrottleDelay", delay);
-		return delayed(file->read(data, length, offset), delay);
+		double diskDelay = getDelay();
+
+		// Wait for diskDelay before submitting the I/O
+		// Template types are being provided explicitly because they can't be automatically deduced for some reason.
+		return mapAsync<Void, std::function<Future<int>(Void)>, int>(
+		    delay(diskDelay), [=](Void _) -> Future<int> { return file->read(data, length, offset); });
 	}
 
 	Future<Void> write(void const* data, int length, int64_t offset) override {
-		double delay = 0.0;
 		char* pdata = nullptr;
+
+		// Check if a bit flip event was injected, if so, copy the buffer contents
+		// with a random bit flipped in a new buffer and use that for the write
 		auto res = g_network->global(INetwork::enBitFlipper);
 		if (res) {
-			auto percentBitFlips = static_cast<BitFlipper*>(res)->getPercentBitFlips();
-			if (percentBitFlips > 0.0) {
-				TraceEvent("AsyncFileCorruptWrite").detail("PercentBitFlips", percentBitFlips);
-				pdata = new char[length];
+			auto bitFlipPercentage = static_cast<BitFlipper*>(res)->getBitFlipPercentage();
+			if (bitFlipPercentage > 0.0) {
+				pdata = (char*)arena.allocate4kAlignedBuffer(length);
 				memcpy(pdata, data, length);
-				flipBits(pdata, length, percentBitFlips);
-				auto diff = memcmp(pdata, data, length);
-				if (diff)
-					TraceEvent("AsyncFileCorruptWriteDiff").detail("Diff", diff);
+				if (deterministicRandom()->random01() < bitFlipPercentage) {
+					// copy buffer with a flipped bit
+					pdata[deterministicRandom()->randomInt(0, length)] ^= (1 << deterministicRandom()->randomInt(0, 8));
+
+					// increment the metric for bit flips
+					auto res = g_network->global(INetwork::enChaosMetrics);
+					if (res) {
+						ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(res);
+						chaosMetrics->bitFlips++;
+					}
+				}
 			}
 		}
 
-		res = g_network->global(INetwork::enDiskFailureInjector);
-		if (res)
-			delay = static_cast<DiskFailureInjector*>(res)->getDiskDelay();
-		TraceEvent("AsyncFileChaosWrite").detail("ThrottleDelay", delay);
-		return delayed(file->write((pdata != nullptr) ? pdata : data, length, offset), delay);
+		double diskDelay = getDelay();
+		// Wait for diskDelay before submitting the I/O
+		return mapAsync<Void, std::function<Future<Void>(Void)>, Void>(delay(diskDelay), [=](Void _) -> Future<Void> {
```

Comment:
`pdata` is just a char pointer, calling `holdWhile()` on it doesn't do anything.  You need to hold the `Arena` that `pdata` is in.